File: python/tests/test_rag.py
"""test_rag.py

Exercises RagContextBuilder + RagPipelineBuilder. Mirrors the TypeScript
rag.test.ts (and CircleAI.Tests.RagContextBuilderTests) plus the fluent-builder
surface and the embedder ranking path.
"""
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Optional

import pytest

from circle_ai.memory.episodic_memory import EpisodicMemoryEntry
from circle_ai.memory.in_memory_episodic_store import InMemoryEpisodicStore
from circle_ai.memory.rag import (
    ITextEmbedder,
    RagContextBuilder,
    RagPipelineBuilder,
)


def _episodic(**overrides) -> EpisodicMemoryEntry:
    return EpisodicMemoryEntry(
        id=overrides.get("id", uuid.uuid4()),
        recorded_at_utc=overrides.get(
            "recorded_at_utc", datetime(2026, 6, 1, 12, 34, tzinfo=timezone.utc)
        ),
        user_text=overrides.get("user_text", "u"),
        assistant_text=overrides.get("assistant_text", "a"),
        app_context=overrides.get("app_context"),
        embedding=overrides.get("embedding"),
        tags=overrides.get("tags"),
    )


def _count_occurrences(text: str, token: str) -> int:
    count = 0
    start = 0
    while True:
        i = text.find(token, start)
        if i < 0:
            break
        count += 1
        start = i + len(token)
    return count


class _FixedEmbedder:
    """Embedder that maps any query to a fixed vector."""

    def __init__(self, vec: list[float]) -> None:
        self._vec = vec

    async def generate_async(
        self, text: str, *, ct: Optional[object] = None
    ) -> list[float]:
        return list(self._vec)


class _ThrowingEmbedder:
    async def generate_async(
        self, text: str, *, ct: Optional[object] = None
    ) -> list[float]:
        raise RuntimeError("embedder offline")


class _ThrowingEpisodicStore:
    """Store that always throws — used to test resilience."""

    async def add_async(self, *a, **k):
        raise RuntimeError("store failure")

    async def search_async(self, *a, **k):
        raise RuntimeError("store failure")

    async def get_recent_async(self, *a, **k):
        raise RuntimeError("store failure")

    async def count_async(self, *a, **k):
        raise RuntimeError("store failure")

    async def prune_older_than_async(self, *a, **k):
        raise RuntimeError("store failure")


# ══════════════════════════════════════════════════════════════════════════
# Constructor guards
# ══════════════════════════════════════════════════════════════════════════


def test_throws_when_store_is_none():
    with pytest.raises(ValueError):
        RagContextBuilder(None)  # type: ignore[arg-type]


# ══════════════════════════════════════════════════════════════════════════
# Empty / missing query
# ══════════════════════════════════════════════════════════════════════════


async def test_empty_query_returns_empty():
    b = RagContextBuilder(InMemoryEpisodicStore())
    assert await b.build_context_async("") == ""


async def test_whitespace_query_returns_empty():
    b = RagContextBuilder(InMemoryEpisodicStore())
    assert await b.build_context_async("   ") == ""


# ══════════════════════════════════════════════════════════════════════════
# Empty store
# ══════════════════════════════════════════════════════════════════════════


async def test_empty_store_returns_empty():
    b = RagContextBuilder(InMemoryEpisodicStore())
    assert await b.build_context_async("hello") == ""


# ══════════════════════════════════════════════════════════════════════════
# Formatting (recency fallback, no embedder)
# ══════════════════════════════════════════════════════════════════════════


async def test_returns_formatted_block_with_header_and_both_texts():
    store = InMemoryEpisodicStore()
    await store.add_async(
        _episodic(
            user_text="What is SDPKT?",
            assistant_text="SDPKT is the TGN wallet.",
            recorded_at_utc=datetime(2026, 6, 1, 11, 0, tzinfo=timezone.utc),
        )
    )

    b = RagContextBuilder(store, None, 3)
    result = await b.build_context_async("tell me about the wallet")

    assert result != ""
    assert "What is SDPKT?" in result
    assert "SDPKT is the TGN wallet." in result
    assert "[Relevant past exchanges" in result


async def test_formats_utc_timestamp_and_labels_user_and_b():
    store = InMemoryEpisodicStore()
    await store.add_async(
        _episodic(
            user_text="q",
            assistant_text="r",
            recorded_at_utc=datetime(2026, 6, 1, 9, 5, tzinfo=timezone.utc),
        )
    )
    b = RagContextBuilder(store, None, 1)
    result = await b.build_context_async("anything")
    assert "[2026-06-01 09:05 UTC]" in result
    assert "User: q" in result
    assert "B!: r" in result


async def test_respects_top_k_counts_bullet_prefixes():
    store = InMemoryEpisodicStore()
    for i in range(10):
        await store.add_async(
            _episodic(user_text=f"question {i}", assistant_text=f"answer {i}")
        )

    b = RagContextBuilder(store, None, 2)
    result = await b.build_context_async("any question")
    assert _count_occurrences(result, "• [") == 2


async def test_includes_app_context_when_set():
    store = InMemoryEpisodicStore()
    await store.add_async(
        _episodic(user_text="bid query", assistant_text="bid answer", app_context="tgn.bidbaas")
    )
    b = RagContextBuilder(store, None, 3)
    result = await b.build_context_async("bidding")
    assert "tgn.bidbaas" in result


async def test_truncates_long_texts_to_half_budget_with_ellipsis():
    store = InMemoryEpisodicStore()
    long_text = "x" * 500
    await store.add_async(_episodic(user_text=long_text, assistant_text="a"))
    # max_chars_per_entry 100 → half 50 → truncate to 49 chars + "…"
    b = RagContextBuilder(store, None, 1, 100)
    result = await b.build_context_async("q")
    assert ("x" * 49 + "…") in result
    assert ("x" * 51) not in result


# ══════════════════════════════════════════════════════════════════════════
# Embedder ranking path
# ══════════════════════════════════════════════════════════════════════════


async def test_ranks_by_embedding_when_embedder_supplied():
    store = InMemoryEpisodicStore()
    await store.add_async(_episodic(user_text="near", assistant_text="n", embedding=[1, 0]))
    await store.add_async(_episodic(user_text="far", assistant_text="f", embedding=[0, 1]))

    b = RagContextBuilder(store, _FixedEmbedder([1, 0]), 1)
    result = await b.build_context_async("anything")
    assert "near" in result
    assert "far" not in result


async def test_falls_back_to_recency_when_embedder_throws():
    store = InMemoryEpisodicStore()
    await store.add_async(
        _episodic(
            user_text="only",
            assistant_text="entry",
            recorded_at_utc=datetime(2026, 6, 1, tzinfo=timezone.utc),
        )
    )
    b = RagContextBuilder(store, _ThrowingEmbedder(), 3)
    result = await b.build_context_async("q")
    assert "only" in result


# ══════════════════════════════════════════════════════════════════════════
# Resilience — store throws
# ══════════════════════════════════════════════════════════════════════════


async def test_returns_empty_when_store_throws():
    b = RagContextBuilder(_ThrowingEpisodicStore())
    assert await b.build_context_async("query") == ""


# ══════════════════════════════════════════════════════════════════════════
# RagPipelineBuilder
# ══════════════════════════════════════════════════════════════════════════


async def test_builder_builds_from_in_memory_store_and_works():
    store = InMemoryEpisodicStore()
    await store.add_async(_episodic(user_text="hi", assistant_text="hello"))
    rag = (
        RagPipelineBuilder.create()
        .with_store(store)
        .with_top_k(2)
        .with_max_chars_per_entry(500)
        .build()
    )
    ctx = await rag.build_context_async("greeting")
    assert "hi" in ctx


async def test_builder_with_in_memory_store_wires_fresh_store():
    rag = RagPipelineBuilder.create().with_in_memory_store().build()
    assert await rag.build_context_async("nothing stored") == ""


def test_builder_build_without_store_throws():
    with pytest.raises(ValueError, match="(?i)episodic memory store is required"):
        RagPipelineBuilder.create().build()


def test_builder_with_top_k_rejects_below_1():
    with pytest.raises(ValueError):
        RagPipelineBuilder.create().with_top_k(0)


def test_builder_with_max_chars_rejects_below_50():
    with pytest.raises(ValueError):
        RagPipelineBuilder.create().with_max_chars_per_entry(49)


async def test_builder_with_embedder_wires_semantic_ranking():
    store = InMemoryEpisodicStore()
    await store.add_async(_episodic(user_text="near", assistant_text="n", embedding=[1, 0]))
    await store.add_async(_episodic(user_text="far", assistant_text="f", embedding=[0, 1]))
    rag = (
        RagPipelineBuilder.create()
        .with_store(store)
        .with_embedder(_FixedEmbedder([1, 0]))
        .with_top_k(1)
        .build()
    )
    ctx = await rag.build_context_async("q")
    assert "near" in ctx


def test_itextembedder_protocol_is_runtime_checkable():
    assert isinstance(_FixedEmbedder([1, 0]), ITextEmbedder)

File: c/tests/test_rag.c
/*
 * test_rag.c — RagContextBuilder + RagPipelineBuilder (C11).
 *
 * Mirrors the verified TypeScript rag.test.ts: empty/blank query, empty store,
 * formatting (UTC timestamp, User/B! labels, app context, truncation), the
 * embedder ranking path, embedder-failure fallback, and the fluent builder.
 *
 * The C# "store throws" resilience test has no C analogue (no exceptions); the
 * embedder-returns-NULL fallback exercises the same degrade-to-recency path.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "circle_ai/circle_ai.h"

/* 2026-06-01T11:00:00Z in Unix ms */
#define MS_2026_06_01_1100  1780311600000LL
/* 2026-06-01T09:05:00Z */
#define MS_2026_06_01_0905  1780304700000LL

static int count_occurrences(const char *text, const char *token) {
    int count = 0;
    const char *p = text;
    size_t tl = strlen(token);
    while ((p = strstr(p, token)) != NULL) { count++; p += tl; }
    return count;
}

/* Embedder that maps any query to the x-axis [1,0]. */
static float *embed_xaxis(void *user, const char *text, size_t *out_len) {
    (void)user; (void)text;
    float *v = (float *)malloc(2 * sizeof(float));
    v[0] = 1.0f; v[1] = 0.0f;
    *out_len = 2;
    return v;
}

/* Embedder that "throws" — returns NULL → recency fallback. */
static float *embed_fail(void *user, const char *text, size_t *out_len) {
    (void)user; (void)text; (void)out_len;
    return NULL;
}

static ca_episodic_entry_t entry(const char *id, int64_t ms, const char *u,
                                 const char *a, const char *ctx,
                                 const float *emb, size_t emblen) {
    ca_episodic_entry_t e = {0};
    e.id = (char *)id;
    e.recorded_at_ms = ms;
    e.user_text = (char *)u;
    e.assistant_text = (char *)a;
    e.app_context = (char *)ctx;
    e.embedding = (float *)emb;
    e.embedding_len = emblen;
    return e;
}

int main(void) {
    /* ── constructor guard ── */
    assert(ca_rag_context_builder_create(NULL, NULL, NULL, 5, 300) == NULL);

    /* ── empty / blank query ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, NULL, NULL, 5, 300);
        char *r1 = ca_rag_context_builder_build(b, "");
        assert(strcmp(r1, "") == 0); free(r1);
        char *r2 = ca_rag_context_builder_build(b, "   ");
        assert(strcmp(r2, "") == 0); free(r2);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── empty store ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, NULL, NULL, 5, 300);
        char *r = ca_rag_context_builder_build(b, "hello");
        assert(strcmp(r, "") == 0); free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── formatting: header + both texts ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        ca_episodic_entry_t e = entry("e1", MS_2026_06_01_1100,
            "What is SDPKT?", "SDPKT is the TGN wallet.", NULL, NULL, 0);
        ca_episodic_store_add(store, &e);
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, NULL, NULL, 3, 300);
        char *r = ca_rag_context_builder_build(b, "tell me about the wallet");
        assert(strcmp(r, "") != 0);
        assert(strstr(r, "What is SDPKT?"));
        assert(strstr(r, "SDPKT is the TGN wallet."));
        assert(strstr(r, "[Relevant past exchanges"));
        free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── formatting: UTC timestamp + User/B! labels ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        ca_episodic_entry_t e = entry("e1", MS_2026_06_01_0905, "q", "r", NULL, NULL, 0);
        ca_episodic_store_add(store, &e);
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, NULL, NULL, 1, 300);
        char *r = ca_rag_context_builder_build(b, "anything");
        assert(strstr(r, "[2026-06-01 09:05 UTC]"));
        assert(strstr(r, "User: q"));
        assert(strstr(r, "B!: r"));
        free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── respects topK (counts bullet prefixes "• [") ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        for (int i = 0; i < 10; ++i) {
            char u[32], a[32], id[16];
            snprintf(u, sizeof(u), "question %d", i);
            snprintf(a, sizeof(a), "answer %d", i);
            snprintf(id, sizeof(id), "e%d", i);
            ca_episodic_entry_t e = entry(id, 1000 + i, u, a, NULL, NULL, 0);
            ca_episodic_store_add(store, &e);
        }
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, NULL, NULL, 2, 300);
        char *r = ca_rag_context_builder_build(b, "any question");
        assert(count_occurrences(r, "\xE2\x80\xA2 [") == 2); /* "• [" */
        free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── includes app context ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        ca_episodic_entry_t e = entry("e1", 1000, "bid query", "bid answer", "tgn.bidbaas", NULL, 0);
        ca_episodic_store_add(store, &e);
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, NULL, NULL, 3, 300);
        char *r = ca_rag_context_builder_build(b, "bidding");
        assert(strstr(r, "tgn.bidbaas"));
        free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── truncates long texts to half-budget with ellipsis ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        char longText[501];
        memset(longText, 'x', 500); longText[500] = '\0';
        ca_episodic_entry_t e = entry("e1", 1000, longText, "a", NULL, NULL, 0);
        ca_episodic_store_add(store, &e);
        /* maxCharsPerEntry 100 → half 50 → truncate to 49 chars + "…" */
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, NULL, NULL, 1, 100);
        char *r = ca_rag_context_builder_build(b, "q");
        char x49[64]; memset(x49, 'x', 49); x49[49] = '\0';
        char needle[80]; snprintf(needle, sizeof(needle), "%s\xE2\x80\xA6", x49); /* 49x + … */
        assert(strstr(r, needle));
        char x51[64]; memset(x51, 'x', 51); x51[51] = '\0';
        assert(strstr(r, x51) == NULL);
        free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── embedder ranking path ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        float near[] = {1.0f, 0.0f};
        float far[]  = {0.0f, 1.0f};
        ca_episodic_entry_t en = entry("near", 1000, "near", "n", NULL, near, 2);
        ca_episodic_entry_t ef = entry("far", 1001, "far", "f", NULL, far, 2);
        ca_episodic_store_add(store, &en);
        ca_episodic_store_add(store, &ef);
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, embed_xaxis, NULL, 1, 300);
        char *r = ca_rag_context_builder_build(b, "anything");
        assert(strstr(r, "near"));
        assert(strstr(r, "far") == NULL);
        free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── embedder throws → recency fallback (still best-effort) ── */
    {
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        ca_episodic_entry_t e = entry("only", 1000, "only", "entry", NULL, NULL, 0);
        ca_episodic_store_add(store, &e);
        ca_rag_context_builder_t *b = ca_rag_context_builder_create(store, embed_fail, NULL, 3, 300);
        char *r = ca_rag_context_builder_build(b, "q");
        assert(strstr(r, "only"));
        free(r);
        ca_rag_context_builder_destroy(b);
        ca_episodic_store_destroy(store);
    }

    /* ── RagPipelineBuilder ── */
    {
        /* build from a store, produce a working builder */
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        ca_episodic_entry_t e = entry("e1", 1000, "hi", "hello", NULL, NULL, 0);
        ca_episodic_store_add(store, &e);
        ca_rag_pipeline_builder_t *pb = ca_rag_pipeline_builder_create();
        assert(ca_rag_pipeline_builder_with_store(pb, store));
        assert(ca_rag_pipeline_builder_with_top_k(pb, 2));
        assert(ca_rag_pipeline_builder_with_max_chars(pb, 500));
        ca_rag_context_builder_t *rag = ca_rag_pipeline_builder_build(pb);
        assert(rag);
        char *ctx = ca_rag_context_builder_build(rag, "greeting");
        assert(strstr(ctx, "hi"));
        free(ctx);
        ca_rag_context_builder_destroy(rag);
        ca_rag_pipeline_builder_destroy(pb);
        ca_episodic_store_destroy(store);
    }
    {
        /* withInMemoryStore wires a fresh (empty) store */
        ca_rag_pipeline_builder_t *pb = ca_rag_pipeline_builder_create();
        assert(ca_rag_pipeline_builder_with_in_memory_store(pb));
        ca_rag_context_builder_t *rag = ca_rag_pipeline_builder_build(pb);
        assert(rag);
        char *ctx = ca_rag_context_builder_build(rag, "nothing stored");
        assert(strcmp(ctx, "") == 0);
        free(ctx);
        ca_rag_context_builder_destroy(rag); /* frees the owned in-memory store */
        ca_rag_pipeline_builder_destroy(pb);
    }
    {
        /* build without a store → NULL */
        ca_rag_pipeline_builder_t *pb = ca_rag_pipeline_builder_create();
        assert(ca_rag_pipeline_builder_build(pb) == NULL);
        ca_rag_pipeline_builder_destroy(pb);
    }
    {
        /* withTopK < 1 rejected, withMaxCharsPerEntry < 50 rejected */
        ca_rag_pipeline_builder_t *pb = ca_rag_pipeline_builder_create();
        assert(ca_rag_pipeline_builder_with_top_k(pb, 0) == NULL);
        assert(ca_rag_pipeline_builder_with_max_chars(pb, 49) == NULL);
        ca_rag_pipeline_builder_destroy(pb);
    }
    {
        /* withEmbedder wires the semantic-ranking seam */
        ca_episodic_store_t *store = ca_episodic_store_create(1024);
        float near[] = {1.0f, 0.0f};
        float far[]  = {0.0f, 1.0f};
        ca_episodic_entry_t en = entry("near", 1000, "near", "n", NULL, near, 2);
        ca_episodic_entry_t ef = entry("far", 1001, "far", "f", NULL, far, 2);
        ca_episodic_store_add(store, &en);
        ca_episodic_store_add(store, &ef);
        ca_rag_pipeline_builder_t *pb = ca_rag_pipeline_builder_create();
        ca_rag_pipeline_builder_with_store(pb, store);
        ca_rag_pipeline_builder_with_embedder(pb, embed_xaxis, NULL);
        ca_rag_pipeline_builder_with_top_k(pb, 1);
        ca_rag_context_builder_t *rag = ca_rag_pipeline_builder_build(pb);
        char *ctx = ca_rag_context_builder_build(rag, "q");
        assert(strstr(ctx, "near"));
        free(ctx);
        ca_rag_context_builder_destroy(rag);
        ca_rag_pipeline_builder_destroy(pb);
        ca_episodic_store_destroy(store);
    }

    printf("test_rag: all assertions passed\n");
    return 0;
}
